src/go/k8s/api/redpanda/v1alpha2/user_types.go

```diff
@@ -153,15 +235,84 @@ type ACLRule struct {
 	Operations []ACLOperation `json:"operations"`
 }
 
+func (r *ACLRule) GetHost() string {
+	return ptr.Deref(r.Host, "*")
+}
+
+func (r *ACLRule) Equals(other ACLRule) bool { //nolint:gocritic // pass by value here is fine
+	if r.Type != other.Type {
+		return false
+	}
+
+	if !r.Resource.Equals(other.Resource) {
+		return false
+	}
+
+	if r.GetHost() != other.GetHost() {
+		return false
+	}
+
+	if len(r.Operations) != len(other.Operations) {
+		return false
+	}
+
+	for _, operation := range r.Operations {
+		if !slices.Contains(other.Operations, operation) {
+			return false
+		}
+	}
+
+	for _, operation := range other.Operations {
+		if !slices.Contains(r.Operations, operation) {
+			return false
+		}
+	}
+
+	return true
+}
+
 // PatternType specifies the type of pattern applied for ACL resource matching.
 // +kubebuilder:validation:Enum=literal;prefixed
 type PatternType string
 
 const (
 	PatternTypeLiteral  PatternType = "literal"
 	PatternTypePrefixed PatternType = "prefixed"
+	PatternTypeUnknown  PatternType = "Unknown"
+)
+
+var (
+	patternTypeFromKafka = map[kmsg.ACLResourcePatternType]PatternType{
+		kmsg.ACLResourcePatternTypeLiteral:  PatternTypeLiteral,
+		kmsg.ACLResourcePatternTypePrefixed: PatternTypePrefixed,
+	}
+	patternTypeToKafka = map[PatternType]kmsg.ACLResourcePatternType{
+		PatternTypeLiteral:  kmsg.ACLResourcePatternTypeLiteral,
+		PatternTypePrefixed: kmsg.ACLResourcePatternTypePrefixed,
+	}
 )
 
+func ACLPatternTypeFromKafka(p kmsg.ACLResourcePatternType) PatternType {
+	if patternType, exists := patternTypeFromKafka[p]; exists {
+		return patternType
+	}
+
+	return PatternTypeUnknown
+}
+
+func (p *PatternType) ToKafka() kmsg.ACLResourcePatternType {
+	if p == nil {
```

Comment:
TBH, this is pretty much just a bit of defensive programming and test ergonomics. It allows me to elide doing an explicit nil check elswhere with what _should be_ otherwise be an unreachable `nil`. `PatternType` defaults to `literal` if unspecified in the CR creation payload, so this just does the same -- if the pointer itself is `nil` just return `literal`. The only time I really imagine this coming up is in tests where I'm passing a directly initialized `ACLRule` without a `PatternType` specified rather than going through the API server which will default the field.

Reply:
Yeah, so this seems to be a funny thing with zero-values. I guess what I'm trying to do with this code is make the parsing behavior roughly equivalent to the json payload behavior so we can just do a simple call chain rather than `nil` check everywhere. With the use of a pointer + a default for the api-server the exact behavior of passing in values is:

1. Pass `nil` --> `literal` because of defaults
2. Pass `"literal"` --> `literal`
3. Pass `"prefixed"` --> `prefixed`
4. Pass `"other"` --> rejected because unknown
5. Pass `""` --> rejected because unknown

The cases in question are 1 and 5. I might be able to adjust the CRD definition such that we're specifying a string rather than a pointer, adding an `omitempty` and still defaulting properly, in which case we'd remove the `nil` condition.

I just dislike the fact that doing so would mean that the struct itself is just relying on the particularities of Go's JSON marshaler to drop zero-values with an `omitempty` tag, such that you'd have to manually construct a json payload to ever pass an empty string directly.

If you still feel strongly about this though, I have no problem changing it in a follow-up.